tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,596 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/kernels/data/dataset_test_base.h"
+
+namespace tensorflow {
+namespace data {
+namespace {
+
+constexpr char kNodeName[] = "concatenate_dataset";
+constexpr char kOpName[] = "ConcatenateDataset";
+
+class ConcatenateDatasetOpTest : public DatasetOpsTestBase {
+ protected:
+  // Creates `TensorSliceDataset` variant tensors from the input vector of
+  // tensor vectors.
+  Status CreateTensorSliceDatasetTensors(
+      const std::vector<std::vector<Tensor>> &tensor_vectors,
+      std::vector<Tensor> *const dataset_tensors) {
+    for (int i = 0; i < tensor_vectors.size(); ++i) {
+      std::vector<Tensor> tensors = tensor_vectors[i];
+      DatasetBase *tensor_slice_dataset;
+      TF_RETURN_IF_ERROR(
+          CreateTensorSliceDataset(strings::StrCat("tensor_slice_node_", i),
+                                   &tensors, &tensor_slice_dataset));
+      Tensor dataset_tensor(DT_VARIANT, TensorShape({}));
+      TF_RETURN_IF_ERROR(
+          StoreDatasetInVariantTensor(tensor_slice_dataset, &dataset_tensor));
+      dataset_tensors->emplace_back(std::move(dataset_tensor));
+    }
+    return Status::OK();
+  }
+
+  // Creates a new ConcatenateDataset op kernel.
+  Status CreateConcatenateDatasetKernel(
+      const DataTypeVector &output_tyeps,
+      const std::vector<PartialTensorShape> &output_shapes,
+      std::unique_ptr<OpKernel> *op_kernel) {
+    node_def_ = test::function::NDef(
+        kNodeName, kOpName, {"input_dataset", "another_dataset"},
+        {{"output_types", output_tyeps}, {"output_shapes", output_shapes}});
+    TF_CHECK_OK(CreateOpKernel(node_def_, op_kernel));
+    return Status::OK();
+  }
+
+  // Creates a new ConcatenateDataset op kernel context.
+  Status CreateConcatenateDatasetContext(
+      OpKernel *const op_kernel,
+      gtl::InlinedVector<TensorValue, 4> *const inputs,
+      std::unique_ptr<OpKernelContext> *context) {
+    TF_RETURN_IF_ERROR(CheckOpKernelInput(*op_kernel, *inputs));
+    TF_RETURN_IF_ERROR(CreateOpKernelContext(op_kernel, inputs, context));
+    return Status::OK();
+  }
+
+ private:
+  NodeDef node_def_;
+};
+
+struct TestParam {
+  std::vector<std::vector<Tensor>> input_tensors;
+  std::vector<Tensor> expected_outputs;
+  DataTypeVector expected_output_dtypes;
+  std::vector<PartialTensorShape> expected_output_shapes;
+  int64 expected_cardinality;
+  std::vector<int> breakpoints;
+} TestCases[] = {
+    // Same shape.
+    {{{DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2, 2}, {1, 2, 3, 4}),
```

Comment:
Add function argument comments to make it clear what these parameters are, e.g.

```suggestion
    {/*input_tensors=*/{{DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2, 2}, {1, 2, 3, 4}),
```

https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

Likewise for the other struct components